Mathlib/Tactic/ComputeDegree.lean

```diff
@@ -4,37 +4,461 @@ Released under Apache 2.0 license as described in the file LICENSE.
+  dbg_trace f!"congr lemma: '{natr}'"
+  nam
+else
+  nam
+
+/--
+`dispatchLemma twoH` takes its input `twoH` from the output of `twoHeadsArgs`.
+
+Using the information contained in `twoH`, it decides which lemma is the most appropriate.
+
+`dispatchLemma` is essentially the main dictionary for `compute_degree`.
+-/
+--  Internally, `dispatchLemma` produces 3 names: these are the lemmas that are appropriate
+--  for goals of the form `natDegree f ≤ d`, `degree f ≤ d`, `coeff f d = a`, in this order.
+def dispatchLemma (twoH : Name × Name × Name × List Bool) (debug : Bool := false) : Name :=
+match twoH with
+  | (.anonymous, _, _, _) => ``id -- `twoH` gave default value, so we do nothing
+  | (_, .anonymous, _, _) => ``id -- `twoH` gave default value, so we do nothing
+  | (na1, na2, head, bools) =>
+    let msg := f!"\ndispatchLemma:\n  {head}"
+    -- if there is some non-metavariable on the way, we "congr" it away
+    if false ∈ bools then getCongrLemma (na1, na2, bools) debug
+    else
+    -- otherwise, we select either the first, second or third element of the triple in `nas` below
+    let π : Name × Name × Name → Name := match na1, na2 with
+      | ``natDegree, ``LE.le => Prod.fst
+      | ``degree, ``LE.le => Prod.fst ∘ Prod.snd
+      | _, _ => Prod.snd ∘ Prod.snd
+    let nas := match head with
+      | `zero => (``natDegree_zero_le, ``degree_zero_le, ``coeff_zero)
+      | `one => (``natDegree_one_le, ``degree_one_le, ``coeff_one)
+      | `many => (``natDegree_nat_cast_le, ``degree_nat_cast_le, ``coeff_nat_cast_ite)
+      | ``HAdd.hAdd =>
+        (``natDegree_add_le_of_le, ``degree_add_le_of_le, ``coeff_add_of_eq)
+      | ``HSub.hSub =>
+        (``natDegree_sub_le_of_le, ``degree_sub_le_of_le, ``coeff_sub_of_eq)
+      | ``HMul.hMul =>
+        (``natDegree_mul_le_of_le, ``degree_mul_le_of_le, ``coeff_mul_add_of_le_natDegree_of_eq_ite)
+      | ``HPow.hPow =>
+        (``natDegree_pow_le_of_le, ``degree_pow_le_of_le, ``coeff_pow_of_natDegree_le_of_eq_ite')
+      | ``Neg.neg =>
+        (``natDegree_neg_le_of_le, ``degree_neg_le_of_le, ``coeff_neg)
+      | ``Polynomial.X =>
+        (``natDegree_X_le, ``degree_X_le, ``coeff_X)
+      | ``Nat.cast =>
+        (``natDegree_nat_cast_le, ``degree_nat_cast_le, ``coeff_nat_cast_ite)
+      | ``NatCast.natCast =>
+        (``natDegree_nat_cast_le, ``degree_nat_cast_le, ``coeff_nat_cast_ite)
+      | ``Int.cast =>
+        (``natDegree_int_cast_le, ``degree_int_cast_le, ``coeff_int_cast_ite)
+      | ``IntCast.intCast =>
+        (``natDegree_int_cast_le, ``degree_int_cast_le, ``coeff_int_cast_ite)
+      | ``Polynomial.monomial =>
+        (``natDegree_monomial_le, ``degree_monomial_le, ``coeff_monomial)
+      | ``Polynomial.C =>
+        (``natDegree_C_le, ``degree_C_le, ``coeff_C)
+      | `error => (.anonymous, .anonymous, .anonymous)
+      | _ => (``le_rfl, ``le_rfl, ``rfl)
+    if debug then dbg_trace f!"{(π nas).getString}\n{msg}"; π nas else π nas
```

Comment:
To make it accept the changes made to `twoHeadArgs`, and some changes to improve robustness.
```suggestion
def dispatchLemma
    (twoH : Name × Name × Sum Name Name × List Bool) (debug : Bool := false) : Name :=
  match twoH with
    | (.anonymous, _, _) => ``id -- `twoH` gave default value, so we do nothing
    | (_, .anonymous, _) => ``id -- `twoH` gave default value, so we do nothing
    | (na1, na2, head, bools) =>
      let msg := f!"\ndispatchLemma:\n  {head}"
      -- if there is some non-metavariable on the way, we "congr" it away
      if false ∈ bools then getCongrLemma (na1, na2, bools) debug
      else
      -- otherwise, we select either the first, second or third element of the triple in `nas` below
      let π (natDegLE : Name) (degLE : Name) (coeff : Name) : Name := Id.run do
        let lem := match na1, na2 with
          | ``natDegree, ``LE.le => natDegLE
          | ``degree, ``LE.le => degLE
          | ``coeff, ``Eq => coeff
          | _, ``LE.le => ``le_rfl
          | _, _ => ``rfl
        if debug then
          dbg_trace f!"{lem.getString}\n{msg}"
        lem
      match head with
        | .inl `zero => π ``natDegree_zero_le ``degree_zero_le ``coeff_zero
        | .inl `one => π ``natDegree_one_le ``degree_one_le ``coeff_one
        | .inl `many => π ``natDegree_nat_cast_le ``degree_nat_cast_le ``coeff_nat_cast_ite
        | .inl .anonymous => π ``le_rfl ``le_rfl ``rfl
        | .inr ``HAdd.hAdd =>
          π ``natDegree_add_le_of_le ``degree_add_le_of_le ``coeff_add_of_eq
        | .inr ``HSub.hSub =>
          π ``natDegree_sub_le_of_le ``degree_sub_le_of_le ``coeff_sub_of_eq
        | .inr ``HMul.hMul =>
          π ``natDegree_mul_le_of_le ``degree_mul_le_of_le ``coeff_mul_add_of_le_natDegree_of_eq_ite
        | .inr ``HPow.hPow =>
          π ``natDegree_pow_le_of_le ``degree_pow_le_of_le ``coeff_pow_of_natDegree_le_of_eq_ite'
        | .inr ``Neg.neg =>
          π ``natDegree_neg_le_of_le ``degree_neg_le_of_le ``coeff_neg
        | .inr ``Polynomial.X =>
          π ``natDegree_X_le ``degree_X_le ``coeff_X
        | .inr ``Nat.cast =>
          π ``natDegree_nat_cast_le ``degree_nat_cast_le ``coeff_nat_cast_ite
        | .inr ``NatCast.natCast =>
          π ``natDegree_nat_cast_le ``degree_nat_cast_le ``coeff_nat_cast_ite
        | .inr ``Int.cast =>
          π ``natDegree_int_cast_le ``degree_int_cast_le ``coeff_int_cast_ite
        | .inr ``IntCast.intCast =>
          π ``natDegree_int_cast_le ``degree_int_cast_le ``coeff_int_cast_ite
        | .inr ``Polynomial.monomial =>
          π ``natDegree_monomial_le ``degree_monomial_le ``coeff_monomial
        | .inr ``Polynomial.C =>
          π ``natDegree_C_le ``degree_C_le ``coeff_C
        | _ => π ``le_rfl ``le_rfl ``rfl
```